osu.Game/Screens/OnlinePlay/DailyChallenge/DailyChallengeTimeRemainingRing.cs

```diff
@@ -0,0 +1,142 @@
+                                {
+                                    Text = "00:00:00",
+                                    Font = OsuFont.TorusAlternate.With(size: 40),
+                                    Anchor = Anchor.Centre,
+                                    Origin = Anchor.Centre,
+                                },
+                                new OsuSpriteText
+                                {
+                                    Text = "remaining",
+                                    Font = OsuFont.Default.With(size: 20),
+                                    Anchor = Anchor.Centre,
+                                    Origin = Anchor.Centre,
+                                }
+                            }
+                        }
+                    }
+                }
+            };
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            StartDate.BindValueChanged(_ => Scheduler.AddOnce(updateState));
+            EndDate.BindValueChanged(_ => Scheduler.AddOnce(updateState));
+            updateState();
+            FinishTransforms(true);
+        }
+
+        private ScheduledDelegate? scheduledUpdate;
+
+        private void updateState()
+        {
+            scheduledUpdate?.Cancel();
+            scheduledUpdate = null;
+
+            const float transition_duration = 300;
+
+            if (StartDate.Value == null || EndDate.Value == null || EndDate.Value < DateTimeOffset.Now)
+            {
+                timeText.Text = TimeSpan.Zero.ToString(@"hh\:mm\:ss");
+                progress.Progress = 0;
+                timeText.FadeColour(colours.Red2, transition_duration, Easing.OutQuint);
+                progress.FadeColour(colours.Red2, transition_duration, Easing.OutQuint);
+                return;
+            }
+
+            var roomDuration = EndDate.Value.Value - StartDate.Value.Value;
+            var remaining = EndDate.Value.Value - DateTimeOffset.Now;
+
+            timeText.Text = remaining.ToString(@"hh\:mm\:ss");
+            progress.Progress = remaining.TotalSeconds / roomDuration.TotalSeconds;
+
+            if (remaining < TimeSpan.FromMinutes(15))
+            {
+                timeText.Colour = progress.Colour = colours.Red1;
+                timeText
+                    .FadeColour(colours.Red1)
+                    .Then().FlashColour(colours.Red0, transition_duration, Easing.OutQuint);
+                progress
+                    .FadeColour(colours.Red1)
+                    .Then().FlashColour(colours.Red0, transition_duration, Easing.OutQuint);
+            }
+            else
+            {
+                timeText.FadeColour(Colour4.White, transition_duration, Easing.OutQuint);
+                progress.FadeColour(colourProvider.Highlight1, transition_duration, Easing.OutQuint);
+            }
+
+            scheduledUpdate = Scheduler.AddDelayed(updateState, 1000);
```

Comment:
This method of updating time every second can occasionally skip updates, but it's probably fine for now.

Reply:
Could also probably make use of the `repeat` parameter, but not sure it's that important.